barretenberg/cpp/src/barretenberg/stdlib/protogalaxy_verifier/protogalaxy_recursive_verifier.cpp

```diff
@@ -1,137 +1,118 @@
+void ProtogalaxyRecursiveVerifier_<DeciderVerificationKeys>::run_oink_verifier_on_one_incomplete_key(
+    const std::shared_ptr<DeciderVK>& key, std::string& domain_separator)
 {
-    domain_separator = domain_separator + "_";
-    OinkVerifier oink_verifier{ builder, inst, transcript, domain_separator };
+    OinkRecursiveVerifier_<Flavor> oink_verifier{ builder, key, transcript, domain_separator + '_' };
     oink_verifier.verify();
 }
 
-// TODO(https://github.com/AztecProtocol/barretenberg/issues/795): The rounds prior to actual verifying are common
-// between decider and folding verifier and could be somehow shared so we do not duplicate code so much.
 template <class DeciderVerificationKeys>
-void ProtogalaxyRecursiveVerifier_<DeciderVerificationKeys>::prepare_for_folding()
+void ProtogalaxyRecursiveVerifier_<DeciderVerificationKeys>::run_oink_verifier_on_each_incomplete_key(
+    const std::vector<FF>& proof)
 {
-    auto index = 0;
-    auto inst = keys_to_fold[0];
+    transcript = std::make_shared<Transcript>(proof);
+    size_t index = 0;
+    auto key = keys_to_fold[0];
     auto domain_separator = std::to_string(index);
-
-    if (!inst->is_accumulator) {
-        receive_and_finalise_key(inst, domain_separator);
-        inst->target_sum = 0;
-        inst->gate_challenges = std::vector<FF>(static_cast<size_t>(inst->verification_key->log_circuit_size), 0);
+    if (!key->is_accumulator) {
+        run_oink_verifier_on_one_incomplete_key(key, domain_separator);
+        key->target_sum = 0;
+        key->gate_challenges = std::vector<FF>(static_cast<size_t>(key->verification_key->log_circuit_size), 0);
     }
     index++;
 
     for (auto it = keys_to_fold.begin() + 1; it != keys_to_fold.end(); it++, index++) {
-        auto inst = *it;
+        auto key = *it;
         auto domain_separator = std::to_string(index);
-        receive_and_finalise_key(inst, domain_separator);
+        run_oink_verifier_on_one_incomplete_key(key, domain_separator);
     }
 }
 
 template <class DeciderVerificationKeys>
 std::shared_ptr<typename DeciderVerificationKeys::DeciderVK> ProtogalaxyRecursiveVerifier_<
     DeciderVerificationKeys>::verify_folding_proof(const StdlibProof<Builder>& proof)
 {
-    using Transcript = typename Flavor::Transcript;
+    static constexpr size_t BATCHED_EXTENDED_LENGTH = DeciderVerificationKeys::BATCHED_EXTENDED_LENGTH;
+    static constexpr size_t NUM_KEYS = DeciderVerificationKeys::NUM;
+    static constexpr size_t COMBINER_LENGTH = BATCHED_EXTENDED_LENGTH - NUM_KEYS;
 
-    transcript = std::make_shared<Transcript>(proof);
-    prepare_for_folding();
+    run_oink_verifier_on_each_incomplete_key(proof);
 
-    auto delta = transcript->template get_challenge<FF>("delta");
-    auto accumulator = get_accumulator();
-    auto deltas =
-        compute_round_challenge_pows(static_cast<size_t>(accumulator->verification_key->log_circuit_size), delta);
+    std::shared_ptr<DeciderVK> accumulator = keys_to_fold[0];
```

Comment:
create a copy of the shared ptr, but the key data is mutated in place. now it's much easier to follow the modifications of the accumulator